// mysql/models/contact_standard_fields.js
const sequelize = require('../factory/sequelize');
const Sequelize = require('sequelize');
const { DataTypes } = Sequelize;


const ContactStandardFields =
    sequelize.define('contact_standard_fields', {
        contactId: {
            type: DataTypes.BIGINT,
            allowNull: false,
            primaryKey: true,
            autoIncrement: true
        },
        fullName: {
            type: DataTypes.STRING(191),
            allowNull: true,
            set(value = '') {
                value = value.substr(0, 191);
                return this.setDataValue('fullName', value.trim());
            }
        },
        firstName: {
            type: DataTypes.STRING(500),
            allowNull: true,
            set(value = '') {
                value = value.substr(0, 500);
                return this.setDataValue('firstName', value.trim());
            }
        },
        lastName: {
            type: DataTypes.STRING(500),
            allowNull: true,
            set(value = '') {
                value = value.substr(0, 500);
                return this.setDataValue('lastName', value.trim());
            }
        },
        profilePic: {
            type: DataTypes.STRING(255),
            allowNull: false,
            defaultValue: ''
        },
        locale: {
            type: DataTypes.STRING(20),
            allowNull: false,
            defaultValue: ''
        },
        countryCode: {
            type: DataTypes.STRING(255),
            allowNull: true
        },
        timezone: {
            type: DataTypes.STRING(20),
            allowNull: false,
            defaultValue: '0'
        },
        email: {
            type: DataTypes.STRING(255),
            allowNull: true
        },
        phone: {
            type: DataTypes.STRING(255),
            allowNull: true
        },
        language: {
            type: DataTypes.STRING(255),
            allowNull: true
        },
        created_at: {
            type: DataTypes.DATE,
            allowNull: true
        },
        updated_at: {
            type: DataTypes.DATE,
            allowNull: true
        },
        deleted_at: {
            type: DataTypes.DATE,
            allowNull: true
        },
        botId: {
            type: DataTypes.BIGINT,
            allowNull: true,
            references: {
                model: 'bot',
                key: 'id'
            }
        }
    }, {
        tableName: 'contact_standard_fields',
        timestamps: true,
        createdAt: 'created_at',
        updatedAt: 'updated_at',
        deletedAt: 'deleted_at',
        underscore: true,
        defaultScope: {
            where: {
                deleted_at: {
                    [Sequelize.Op.eq]: null
                }
            }
        }
    });

ContactStandardFields.updateProfilePic = async function(contactId,profilePic) {
    return await ContactStandardFields.update({
        profilePic
    }, { where: { contactId: contactId } });
}
module.exports = ContactStandardFields;
