// mysql/models/contact.js
const sequelize = require('../factory/sequelize');
const Sequelize = require('sequelize');
const {DataTypes} = Sequelize;
const ContactStandardFields = require('./contact_standard_fields');

const Contact = sequelize.define('contact', {
    id: {
        type: DataTypes.BIGINT,
        allowNull: false,
        primaryKey: true,
        autoIncrement: true
    },
    primaryId: {
        type: DataTypes.BIGINT,
        allowNull: true
    },
    botStatus: {
        type: DataTypes.INTEGER(1),
        allowNull: false,
        defaultValue: '1'
    },
    status: {
        type: DataTypes.ENUM('done', 'pending', 'snoozed'),
        allowNull: false,
        defaultValue: 'pending'
    },
    assigneeUserId: {
        type: DataTypes.BIGINT,
        allowNull: true,
        references: {
            model: 'user',
            key: 'id'
        }
    },
    botId: {
        type: DataTypes.BIGINT,
        allowNull: false,
        references: {
            model: 'bot',
            key: 'id'
        }
    },
    statusChangedSince: {
        type: DataTypes.BIGINT,
        allowNull: false
    },
    created_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    updated_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    deleted_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    lastInteractionTime: {
        type: DataTypes.BIGINT,
        allowNull: false,
        defaultValue: '0'
    },
    totalMessageCount: {
        type: DataTypes.INTEGER(11),
        allowNull: true
    }
}, {
    tableName: 'contact',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    deletedAt: 'deleted_at',
    underscore: true,
    defaultScope: {
        where: {
            deleted_at: {
                [Sequelize.Op.eq]: null
            }
        }
    }
});

ContactStandardFields.belongsTo(Contact);
Contact.hasOne(ContactStandardFields);
Contact.getContact = async function (contactId) {
    return await Contact.findOne({
        where: {
            id: contactId
        },
        include: [
            {
                model: ContactStandardFields,
                attributes: {exclude: ['contactId']},
                required: true
            },
        ]
    });
}
module.exports = Contact;
